numpy/ma/core.py

```diff
@@ -2785,10 +2785,12 @@ def __new__(cls, data=None, mask=nomask, dtype=None, copy=False,
         if isinstance(data, MaskedArray) and (data.shape != _data.shape):
             copy = True
         # Careful, cls might not always be MaskedArray.
-        if not isinstance(data, cls) or not subok:
-            _data = ndarray.view(_data, cls)
-        else:
+        # we must never do .view(MaskedConstant), as that would create a new
+        # instance of np.ma.masked, which would be confusing
+        if isinstance(data, cls) and subok and not isinstance(data, MaskedConstant):
             _data = ndarray.view(_data, type(data))
```

Comment:
I get so confused about this: why exactly is a view taken here? For `isinstance(data, cls)` to hold, `data` must be a subclass of `ndarray` as well, and if `subok` was true, then the above assignment of `_data = np.array(data, subok=sub)` would already have ensured that `type(_data) is type(data)` (independently of the other arguments). Am I missing something? If so, do add that as a comment....